cmd/boulder/main.go

```diff
@@ -60,57 +56,89 @@ func readAndValidateConfigFile(name, filename string) error {
+func getConfigPath() string {
+	for i := 0; i < len(os.Args); i++ {
+		arg := os.Args[i]
+		if arg == "--config" || arg == "-config" {
+			if i+1 < len(os.Args) {
+				return os.Args[i+1]
+			}
+		}
+		if strings.HasPrefix(arg, "--config=") {
+			return strings.TrimPrefix(arg, "--config=")
+		}
+		if strings.HasPrefix(arg, "-config=") {
+			return strings.TrimPrefix(arg, "-config=")
+		}
+	}
+	return ""
 }
 
-func init() {
-	cmd.RegisterCommand("boulder", func() {
+var boulderUsage = fmt.Sprintf(`Usage: %s <subcommand> [flags]
+
+  Each boulder component has its own subcommand. Use --list to see
+  a list of the available components. Use <subcommand> --help to
+  see the usage for a specific component.
+`,
+	core.Command())
+
+func main() {
+	var subcommand func()
+	if core.Command() == "boulder" {
+		// The operator is running the boulder binary directly.
 		if len(os.Args) <= 1 {
-			fmt.Fprintf(os.Stderr, "Call with --list to list available subcommands. Run them like boulder <subcommand>.\n")
+			// No arguments passed.
+			fmt.Fprint(os.Stderr, boulderUsage)
 			return
 		}
-		subcommand := cmd.LookupCommand(os.Args[1])
-		if subcommand == nil {
-			fmt.Fprintf(os.Stderr, "Unknown subcommand '%s'.\n", os.Args[1])
+
+		if os.Args[1] == "--help" || os.Args[1] == "-help" {
+			// Help flag passed.
+			fmt.Fprint(os.Stderr, boulderUsage)
 			return
 		}
-		os.Args = os.Args[1:]
-		subcommand()
-	}, nil)
-	// TODO(#6763): Move this inside of main().
-	cmd.RegisterCommand("--list", func() {
-		for _, c := range cmd.AvailableCommands() {
-			if c != "boulder" && c != "--list" {
-				fmt.Println(c)
-			}
-		}
-	}, nil)
-	// TODO(#6763): Move this inside of main().
-	cmd.RegisterCommand("validate", func() {
-		if len(os.Args) <= 1 {
-			fmt.Fprintf(os.Stderr, "Call with --help to list usage.\n")
-			os.Exit(1)
-		}
-		list := flag.Bool("list", false, "List available components to validate configuration for.")
-		component := flag.String("component", "", "The name of the component to validate configuration for.")
-		configFile := flag.String("config", "", "The path to the configuration file to validate.")
-		flag.Parse()
 
-		if *list {
-			for _, c := range cmd.AvailableConfigValidators() {
+		if os.Args[1] == "--list" || os.Args[1] == "-list" {
+			// List flag passed.
+			for _, c := range cmd.AvailableCommands() {
 				fmt.Println(c)
 			}
 			return
 		}
-		if *component == "" || *configFile == "" {
-			fmt.Fprintf(os.Stderr, "Must provide a configuration file to validate.\n")
+
+		// Operator ran a boulder using a subcommand. Lookup the
+		// subcommand.
+		subcommand = cmd.LookupCommand(os.Args[1])
+		if subcommand == nil {
+			fmt.Fprintf(os.Stderr, "Unknown subcommand %q.\n", os.Args[1])
 			os.Exit(1)
 		}
-		err := readAndValidateConfigFile(*component, *configFile)
+	}
+
+	configPath := getConfigPath()
+	if configPath != "" {
+		// Config flag passed.
+		err := readAndValidateConfigFile(core.Command(), configPath)
```

Comment:
If `core.Command()` is "boulder" (i.e. we just executed the if clause above), isn't this going to fail to find a config struct to validate against?

Reply:
😅 yep, I used to assign `os.Args = os.Args[1:]` above this line but moved it recently and now it's definitely broken. I'll refactor this.